### ISSUE TITLE
Error computing output names and sen2r shuts down.

### ISSUE THREAD

**Issue by srahman16** [ranghetti/sen2r#185]
Hi Luigi,

I can't read all the Sentinel-2 (L1C) scenes. Sen2r shuts down after showing these messages,
```
[2019-05-13 01:50:49] Gui closed by the user. Starting processing.
[2019-05-13 01:50:49] Searching for available SAFE products on SciHub...
[2019-05-13 01:53:58] Computing output names...
Error in `[.data.table`(s2_dt, as.Date(gsub("^S2[AB]\\_MSIL[12][AC]\\_([0-9]{8})T.+$",  : 
  i evaluates to a logical vector length 17 but there are 109 rows. Recycling of logical i is no longer allowed as it hides more bugs than is worth the rare convenience. Explicitly use rep(...,length=.N) if you really need to recycle.
```
Great if you can help in suggesting how to solve this issue.

Cheers,

S

**Comment by ranghetti** [comment]
Hi,
in order to reproduce the error, can you post the JSON parameter file of the processing which you are launching?

**Comment by srahman16** [comment]
{
  "preprocess": [true],
  "s2_levels": ["l1c", "l2a"],
  "sel_sensor": ["s2a", "s2b"],
  "online": [true],
  "downloader": ["wget"],
  "overwrite_safe": [false],
  "rm_safe": ["l1c"],
  "max_cloud_safe": [5],
  "step_atmcorr": ["auto"],
  "timewindow": ["2017-01-01", "2019-04-30"],
  "timeperiod": ["full"],
  "extent": ["{\n  \"type\": \"FeatureCollection\",\n  \"features\": [\n    {\n      \"type\": \"Feature\",\n      \"properties\": {\n        \"_leaflet_id\": 526,\n        \"feature_type\": \"polygon\"\n      },\n      \"geometry\": {\n        \"type\": \"Polygon\",\n        \"coordinates\": [\n          [\n            [150.91713, -33.930827],\n            [150.770187, -34.108393],\n            [150.988541, -34.250406],\n            [151.058578, -34.195901],\n            [151.123123, -34.147044],\n            [151.149216, -34.125448],\n            [151.187668, -34.075412],\n            [151.172562, -34.058348],\n            [151.127243, -34.058348],\n            [151.090164, -34.037867],\n            [151.036606, -33.985502],\n            [150.973434, -33.935385],\n            [150.937729, -33.921711],\n            [150.91713, -33.930827]\n          ]\n        ]\n      }\n    }\n  ]\n}"],
  "s2tiles_selected": ["56HKH", "56HLH"],
  "s2orbits_selected": [null],
  "list_prods": ["BOA"],
  "list_indices": ["EVI2", "MIRBI", "NBR2", "NDRE", "NDVI"],
  "list_rgb": [null],
  "rgb_ranges": [null],
  "index_source": ["BOA"],
  "mask_type": [null],
  "max_mask": [80],
  "mask_smooth": [0],
  "mask_buffer": [0],
  "clip_on_extent": [true],
  "extent_as_mask": [true],
  "extent_name": ["sen2r"],
  "reference_path": [null],
  "res": [null],
  "res_s2": ["10m"],
  "unit": ["Meter"],
  "proj": ["+proj=utm +zone=56 +south +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs"],
  "resampling": ["near"],
  "resampling_scl": ["near"],
  "outformat": ["GTiff"],
  "rgb_outformat": ["GTiff"],
  "index_datatype": ["Int16"],
  "compression": ["DEFLATE"],
  "rgb_compression": ["DEFLATE"],
  "overwrite": [false],
  "path_l1c": ["C:/sen2r"],
  "path_l2a": ["C:/sen2r/processed"],
  "path_tiles": [null],
  "path_merged": [null],
  "path_out": ["C:/sen2r/indices"],
  "path_rgb": [""],
  "path_indices": [""],
  "path_subdirs": [true],
  "thumbnails": [true],
  "log": [null],
  "parallel": [true],
  "processing_order": ["by_groups"],
  "pkg_version": ["1.0.2"]
}

**Comment by ranghetti** [comment]
Thank you to report it, it should be fixed now; can you check it and let me know?

**Comment by srahman16** [comment]
I will check and inform it to you.

**Comment by srahman16** [comment]
Sorry, Luigi. No success. Still it shows the same thing and shuts down.

**Comment by ranghetti** [comment]
Have you reinstalled the package?

**Comment by srahman16** [comment]
Yes, Luigi. I am doing it once again and will restart my PC. I will inform you if there will be an issue.

**Comment by srahman16** [comment]
> sen2r() [2019-05-13 18:11:39] Starting sen2r execution. 
[2019-05-13 18:11:40] Launching GUI...  Listening on http://127.0.0.1:5636 
[2019-05-13 18:12:17] Gui closed by the user. Starting processing. 
[2019-05-13 18:12:17] Searching for available SAFE products on SciHub... 
[2019-05-13 18:15:13] Computing output names... Error in `[.data.table`(s2_dt, as.Date(gsub("^S2[AB]\\_MSIL[12][AC]\\_([0-9]{8})T.+$",  :    i evaluates to a logical vector length 17 but there are 109 rows. Recycling of logical i is no longer allowed as it hides more bugs than is worth the rare convenience. Explicitly use rep(...,length=.N) if you really need to recycle.
--
Same report. Can you share the consolidated R script, so I can run it step-by-step?

Regards,

S

**Comment by ranghetti** [comment]
The commit which should heve solved is this: 8e8a7b224bb81cf3db5d6de60cdfca7352a2c568.
It is strange that the same error is still present, since, after the commit, the logical vector is now length 109.
If you want to debug the package (if I understood your request), you can browser function [`sen2r()`](https://github.com/ranghetti/sen2r/blob/master/R/sen2r.R) at line 1398.

**Comment by srahman16** [comment]
The code terminates while getting the logical vector. Somehow, it can't get the length of the logical vector. I am wondering if you have any alternative suggestions to solve this issue.

**Comment by ranghetti** [comment]
If you have browsered the function, before launching the command which returns the error, can you launch
```r
as.Date(
  gsub("^S2[AB]\\_MSIL[12][AC]\\_([0-9]{8})T.+$", "\\1", name),
  "%Y%m%d"
)
nrow(s2_dt)
```
?

**Comment by ranghetti** [comment]
[I updated the instructions in the comment above]

**Comment by srahman16** [comment]
Thank you, Luigi. Sorry but nothing changes, can you send me the R script which is working on your side?

**Comment by ranghetti** [comment]
The R script which is working on my side is the function `sen2r()` updated with commit 8e8a7b224bb81cf3db5d6de60cdfca7352a2c568, that is the same you can find online. I simply edited that function, pushed the edit online and launched `sen2r()` with the JSON you provided. Before the edit I saw the same error you posted, after the edit the error was solved.

In the previous comment I asked you to run two R commands and post me the R output in order to understand the problem, assuming that you were browsering the source code (I interpreted your request of the code in this way). If you are not browsering the source code, I cannot help you in that way. In this case, my impression is that the old version on the package is still installed somewhere. So, try in this way:

1. find the location of the package:
    ```r
    ( sen2r_location <- system.file(package="sen2r") ) 
    ```
2. remove the package installed there:
    ```r
    remove.packages("sen2r", lib=dirname(sen2r_location))
    ```
3. run the command above once again, ensuring that it returns the following error:
   ```
    Error in remove.packages : there is no package called ‘sen2r’
    ```
4. reinstall the package in the same location:
    ```r
    whitr::with_libpaths(new = dirname(sen2r_location), remotes::install_github("ranghetti/sen2r"))
    ```
5. try again.

Please post the content of `sen2r_location` in your next message.

**Comment by srahman16** [comment]
Hi Luigi,

Thank you very much for your help and support. The script is working now,
.........................................................................................
[2019-05-13 21:13:10] Gui closed by the user. Starting processing.
[2019-05-13 21:13:10] Searching for available SAFE products on SciHub...
[2019-05-13 21:15:59] Computing output names...
[2019-05-13 21:24:20] Processing group 1 of 28...
[2019-05-13 21:24:20] Starting to download the required level-2A SAFE products.
[2019-05-13 21:24:21] Download of level-2A SAFE products terminated.
[2019-05-13 21:24:21] Starting to download the required level-1C SAFE products.
[2019-05-13 21:24:23] Downloading (if missing) product 1 of 1 (S2A_MSIL1C_20180927T000241_N0206_R030_T56HLH_20180927T014413.SAFE)...
...........................................................................................

What I did, I reinstall R and also get rid of sen2r as per your suggestions. I am using R version 3.6.0, so "whitr" package is not available for R (3.5.2 and 3.6.0). But, it works fine, and this is the location where sen2cor is installed,

C:\Users\*******\Documents\R\win-library\3.6\sen2r\sen2cor\Sen2Cor-02.05.05-win64

Regards,

S
